Build and link errors with ittnotify enabled

## Expected Behavior

I tried to build hpx with the flags -DHPX_WITH_ITTNOTIFY=ON -DAMPLIFIER_ROOT=/opt/intel-ittapi/inst

I used https://github.com/intel/ittapi for the amplifier.

## Actual Behavior and Steps to Reproduce the Problem

1. missing include error -> it seems the cmake environment doesn't sets the include path
```zsh
/opt/hpx/src/hpx-1.6.0/libs/core/itt_notify/src/itt_notify.cpp:15:10: fatal error: ittnotify.h: No such file or directory
   15 | #include <ittnotify.h>
```
   -> solved by setting CPATH

2. compile error:
```zsh
   /opt/hpx/src/hpx-1.6.0/libs/core/itt_notify/src/itt_notify.cpp: In constructor 'hpx::util::itt::thread_domain::thread_domain()':
   /opt/hpx/src/hpx-1.6.0/libs/core/itt_notify/src/itt_notify.cpp:267:9: error: class 'hpx::util::itt::thread_domain' does not have any field  named 'domain_'
267 |       : domain_()
```
   -> fixed by replacing : domain_() with domain()   //calling now default constructor instead of base class member

3. link error Linking CXX shared library ../lib/libhpx_core.so , same problem as 1.
   -> fixed with manual linking

4. compile error: 
```zsh
In file included from /opt/hpx/src/hpx-1.6.0/libs/full/actions_base/include/hpx/actions_base/component_action.hpp:12,
                 from /opt/hpx/src/hpx-1.6.0/libs/full/agas/include/hpx/agas/server/primary_namespace.hpp:16,
                 from /opt/hpx/src/hpx-1.6.0/libs/full/agas/src/primary_namespace.cpp:12:
/opt/hpx/src/hpx-1.6.0/libs/full/actions_base/include/hpx/actions_base/basic_action.hpp:49:10: fatal error: hpx/concurrency/itt_notify.hpp: No such file or directory
   49 | #include <hpx/concurrency/itt_notify.hpp>
```

## Specifications

... Please describe your environment

  - HPX Version: 1.6.0 and current master 09ec51f07a59c89b614806f9ceee17d057cb35e7
  - Platform (compiler, OS): Arch Linux
  - Compiler: gcc 10.2.0

--- a/libs/CMakeLists.txt
+++ b/libs/CMakeLists.txt
@@ -114,3 +114,7 @@ endif()
 if(MSVC AND HPX_COROUTINES_WITH_SWAP_CONTEXT_EMULATION)
   target_link_options(hpx_core PRIVATE "/EXPORT:switch_to_fiber")
 endif()
+
+if(HPX_WITH_ITTNOTIFY)
+  target_link_libraries(hpx_core PUBLIC Amplifier::amplifier)
+endif()

--- a/libs/core/hashing/include/hpx/hashing/fibhash.hpp
+++ b/libs/core/hashing/include/hpx/hashing/fibhash.hpp
@@ -24,25 +24,25 @@ namespace hpx { namespace util {
         template <>
         struct hash_helper<0>
         {
-            HPX_STATIC_CONSTEXPR int log2 = -1;
+            static constexpr int log2 = -1;
         };
 
         template <std::uint64_t N>
         struct hash_helper
         {
-            HPX_STATIC_CONSTEXPR std::uint64_t log2 =
+            static constexpr std::uint64_t log2 =
                 hash_helper<(N >> 1)>::log2 + 1;
-            HPX_STATIC_CONSTEXPR std::uint64_t shift_amount = 64 - log2;
+            static constexpr std::uint64_t shift_amount = 64 - log2;
         };
 
-        HPX_STATIC_CONSTEXPR std::uint64_t golden_ratio =
+        static HPX_INLINE_CONSTEXPR_VARIABLE std::uint64_t golden_ratio =
             11400714819323198485llu;
     }    // namespace detail
 
     // This function calculates the hash based on a multiplicative Fibonacci
     // scheme
     template <std::uint64_t N>
-    constexpr std::uint64_t fibhash(std::uint64_t i)
+    constexpr std::uint64_t fibhash(std::uint64_t i) noexcept
     {
         using helper = detail::hash_helper<N>;
         static_assert(N != 0, "This algorithm only works with N != 0");

--- a/libs/core/itt_notify/CMakeLists.txt
+++ b/libs/core/itt_notify/CMakeLists.txt
@@ -1,4 +1,4 @@
-# Copyright (c) 2019-2020 The STE||AR-Group
+# Copyright (c) 2019-2021 The STE||AR-Group
 #
 # SPDX-License-Identifier: BSL-1.0
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
@@ -20,13 +20,17 @@ set(itt_notify_compat_headers
 
 set(itt_notify_sources itt_notify.cpp thread_name.cpp)
 
+if(HPX_WITH_ITTNOTIFY)
+  set(itt_notify_dependencies DEPENDENCIES Amplifier::amplifier)
+endif()
+
 include(HPX_AddModule)
 add_hpx_module(
   core itt_notify
   GLOBAL_HEADER_GEN OFF
   SOURCES ${itt_notify_sources}
   HEADERS ${itt_notify_headers}
-  COMPAT_HEADERS ${itt_notify_compat_headers}
+  COMPAT_HEADERS ${itt_notify_compat_headers} ${itt_notify_dependencies}
   MODULE_DEPENDENCIES hpx_config
   CMAKE_SUBDIRS examples tests
 )

--- a/libs/core/itt_notify/CMakeLists.txt
+++ b/libs/core/itt_notify/CMakeLists.txt
@@ -21,7 +21,7 @@ set(itt_notify_compat_headers
 set(itt_notify_sources itt_notify.cpp thread_name.cpp)
 
 if(HPX_WITH_ITTNOTIFY)
-  set(itt_notify_dependencies DEPENDENCIES Amplifier::amplifier)
+  set(itt_notify_dependencies Amplifier::amplifier)
 endif()
 
 include(HPX_AddModule)
@@ -30,7 +30,8 @@ add_hpx_module(
   GLOBAL_HEADER_GEN OFF
   SOURCES ${itt_notify_sources}
   HEADERS ${itt_notify_headers}
-  COMPAT_HEADERS ${itt_notify_compat_headers} ${itt_notify_dependencies}
+  COMPAT_HEADERS ${itt_notify_compat_headers}
+  DEPENDENCIES ${itt_notify_dependencies}
   MODULE_DEPENDENCIES hpx_config
   CMAKE_SUBDIRS examples tests
 )

--- a/libs/core/itt_notify/include/hpx/modules/itt_notify.hpp
+++ b/libs/core/itt_notify/include/hpx/modules/itt_notify.hpp
@@ -1,4 +1,4 @@
-//  Copyright (c) 2007-2017 Hartmut Kaiser
+//  Copyright (c) 2007-2021 Hartmut Kaiser
 //
 //  SPDX-License-Identifier: BSL-1.0
 //  Distributed under the Boost Software License, Version 1.0. (See accompanying
@@ -16,7 +16,7 @@ struct ___itt_caller;
 struct ___itt_string_handle;
 struct ___itt_domain;
 struct ___itt_id;
-typedef void* __itt_heap_function;
+using __itt_heap_function = void*;
 struct ___itt_counter;
 
 ///////////////////////////////////////////////////////////////////////////////
@@ -85,91 +85,98 @@ struct ___itt_counter;
 // decide whether to use the ITT notify API if it's available
 
 #if HPX_HAVE_ITTNOTIFY != 0
-extern bool use_ittnotify_api;
+HPX_CORE_EXPORT extern bool use_ittnotify_api;
 
 ///////////////////////////////////////////////////////////////////////////////
 HPX_CORE_EXPORT void itt_sync_create(
-    void* addr, const char* objtype, const char* objname);
-HPX_CORE_EXPORT void itt_sync_rename(void* addr, const char* name);
-HPX_CORE_EXPORT void itt_sync_prepare(void* addr);
-HPX_CORE_EXPORT void itt_sync_acquired(void* addr);
-HPX_CORE_EXPORT void itt_sync_cancel(void* addr);
-HPX_CORE_EXPORT void itt_sync_releasing(void* addr);
-HPX_CORE_EXPORT void itt_sync_released(void* addr);
-HPX_CORE_EXPORT void itt_sync_destroy(void* addr);
+    void* addr, const char* objtype, const char* objname) noexcept;
+HPX_CORE_EXPORT void itt_sync_rename(void* addr, const char* name) noexcept;
+HPX_CORE_EXPORT void itt_sync_prepare(void* addr) noexcept;
+HPX_CORE_EXPORT void itt_sync_acquired(void* addr) noexcept;
+HPX_CORE_EXPORT void itt_sync_cancel(void* addr) noexcept;
+HPX_CORE_EXPORT void itt_sync_releasing(void* addr) noexcept;
+HPX_CORE_EXPORT void itt_sync_released(void* addr) noexcept;
+HPX_CORE_EXPORT void itt_sync_destroy(void* addr) noexcept;
 
-HPX_CORE_EXPORT ___itt_caller* itt_stack_create();
-HPX_CORE_EXPORT void itt_stack_enter(___itt_caller* ctx);
-HPX_CORE_EXPORT void itt_stack_leave(___itt_caller* ctx);
-HPX_CORE_EXPORT void itt_stack_destroy(___itt_caller* ctx);
+HPX_CORE_EXPORT ___itt_caller* itt_stack_create() noexcept;
+HPX_CORE_EXPORT void itt_stack_enter(___itt_caller* ctx) noexcept;
+HPX_CORE_EXPORT void itt_stack_leave(___itt_caller* ctx) noexcept;
+HPX_CORE_EXPORT void itt_stack_destroy(___itt_caller* ctx) noexcept;
 
-HPX_CORE_EXPORT void itt_frame_begin(___itt_domain const* frame, ___itt_id* id);
-HPX_CORE_EXPORT void itt_frame_end(___itt_domain const* frame, ___itt_id* id);
+HPX_CORE_EXPORT void itt_frame_begin(
+    ___itt_domain const* frame, ___itt_id* id) noexcept;
+HPX_CORE_EXPORT void itt_frame_end(
+    ___itt_domain const* frame, ___itt_id* id) noexcept;
 
-HPX_CORE_EXPORT int itt_mark_create(char const*);
-HPX_CORE_EXPORT void itt_mark_off(int mark);
-HPX_CORE_EXPORT void itt_mark(int mark, char const*);
+HPX_CORE_EXPORT int itt_mark_create(char const*) noexcept;
+HPX_CORE_EXPORT void itt_mark_off(int mark) noexcept;
+HPX_CORE_EXPORT void itt_mark(int mark, char const*) noexcept;
 
-HPX_CORE_EXPORT void itt_thread_set_name(char const*);
-HPX_CORE_EXPORT void itt_thread_ignore();
+HPX_CORE_EXPORT void itt_thread_set_name(char const*) noexcept;
+HPX_CORE_EXPORT void itt_thread_ignore() noexcept;
 
 HPX_CORE_EXPORT void itt_task_begin(
-    ___itt_domain const*, ___itt_string_handle*);
+    ___itt_domain const*, ___itt_string_handle*) noexcept;
 HPX_CORE_EXPORT void itt_task_begin(
-    ___itt_domain const*, ___itt_id*, ___itt_string_handle*);
-HPX_CORE_EXPORT void itt_task_end(___itt_domain const*);
+    ___itt_domain const*, ___itt_id*, ___itt_string_handle*) noexcept;
+HPX_CORE_EXPORT void itt_task_end(___itt_domain const*) noexcept;
 
-HPX_CORE_EXPORT ___itt_domain* itt_domain_create(char const*);
-HPX_CORE_EXPORT ___itt_string_handle* itt_string_handle_create(char const*);
+HPX_CORE_EXPORT ___itt_domain* itt_domain_create(char const*) noexcept;
+HPX_CORE_EXPORT ___itt_string_handle* itt_string_handle_create(
+    char const*) noexcept;
 
 HPX_CORE_EXPORT ___itt_id* itt_make_id(void*, std::size_t);
-HPX_CORE_EXPORT void itt_id_create(___itt_domain const*, ___itt_id* id);
-HPX_CORE_EXPORT void itt_id_destroy(___itt_id* id);
+HPX_CORE_EXPORT void itt_id_create(
+    ___itt_domain const*, ___itt_id* id) noexcept;
+HPX_CORE_EXPORT void itt_id_destroy(___itt_id* id) noexcept;
 
 HPX_CORE_EXPORT __itt_heap_function itt_heap_function_create(
-    const char*, const char*);
+    const char*, const char*) noexcept;
 HPX_CORE_EXPORT void itt_heap_allocate_begin(
-    __itt_heap_function, std::size_t, int);
+    __itt_heap_function, std::size_t, int) noexcept;
 HPX_CORE_EXPORT void itt_heap_allocate_end(
-    __itt_heap_function, void**, std::size_t, int);
-HPX_CORE_EXPORT void itt_heap_free_begin(__itt_heap_function, void*);
-HPX_CORE_EXPORT void itt_heap_free_end(__itt_heap_function, void*);
+    __itt_heap_function, void**, std::size_t, int) noexcept;
+HPX_CORE_EXPORT void itt_heap_free_begin(__itt_heap_function, void*) noexcept;
+HPX_CORE_EXPORT void itt_heap_free_end(__itt_heap_function, void*) noexcept;
 HPX_CORE_EXPORT void itt_heap_reallocate_begin(
-    __itt_heap_function, void*, std::size_t, int);
+    __itt_heap_function, void*, std::size_t, int) noexcept;
 HPX_CORE_EXPORT void itt_heap_reallocate_end(
-    __itt_heap_function, void*, void**, std::size_t, int);
-HPX_CORE_EXPORT void itt_heap_internal_access_begin();
-HPX_CORE_EXPORT void itt_heap_internal_access_end();
+    __itt_heap_function, void*, void**, std::size_t, int) noexcept;
+HPX_CORE_EXPORT void itt_heap_internal_access_begin() noexcept;
+HPX_CORE_EXPORT void itt_heap_internal_access_end() noexcept;
 
-HPX_CORE_EXPORT ___itt_counter* itt_counter_create(char const*, char const*);
+HPX_CORE_EXPORT ___itt_counter* itt_counter_create(
+    char const*, char const*) noexcept;
 HPX_CORE_EXPORT ___itt_counter* itt_counter_create_typed(
-    char const*, char const*, int);
-HPX_CORE_EXPORT void itt_counter_destroy(___itt_counter*);
-HPX_CORE_EXPORT void itt_counter_set_value(___itt_counter*, void*);
+    char const*, char const*, int) noexcept;
+HPX_CORE_EXPORT void itt_counter_destroy(___itt_counter*) noexcept;
+HPX_CORE_EXPORT void itt_counter_set_value(___itt_counter*, void*) noexcept;
 
-HPX_CORE_EXPORT int itt_event_create(char const* name, int namelen);
-HPX_CORE_EXPORT int itt_event_start(int evnt);
-HPX_CORE_EXPORT int itt_event_end(int evnt);
+HPX_CORE_EXPORT int itt_event_create(char const* name, int namelen) noexcept;
+HPX_CORE_EXPORT int itt_event_start(int evnt) noexcept;
+HPX_CORE_EXPORT int itt_event_end(int evnt) noexcept;
 
 HPX_CORE_EXPORT void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, std::uint64_t const& data);
+    ___itt_string_handle* key, std::uint64_t const& data) noexcept;
 HPX_CORE_EXPORT void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, double const& data);
+    ___itt_string_handle* key, double const& data) noexcept;
 HPX_CORE_EXPORT void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, char const* data);
+    ___itt_string_handle* key, char const* data) noexcept;
 HPX_CORE_EXPORT void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, void const* data);
+    ___itt_string_handle* key, void const* data) noexcept;
 
 ///////////////////////////////////////////////////////////////////////////////
 namespace hpx { namespace util {
+
     struct thread_description;
 }}    // namespace hpx::util
 
 namespace hpx { namespace util { namespace itt {
+
     struct stack_context
     {
         stack_context()
-          : itt_context_(0)
+          : itt_context_(nullptr)
         {
             HPX_ITT_STACK_CREATE(itt_context_);
         }
@@ -220,8 +227,8 @@ namespace hpx { namespace util { namespace itt {
     {
         HPX_NON_COPYABLE(domain);
 
-        HPX_CORE_EXPORT domain(char const*);
-        HPX_CORE_EXPORT domain();
+        HPX_CORE_EXPORT domain(char const*) noexcept;
+        HPX_CORE_EXPORT domain() noexcept;
 
         ___itt_domain* domain_;
     };
@@ -230,12 +237,12 @@ namespace hpx { namespace util { namespace itt {
     {
         HPX_NON_COPYABLE(thread_domain);
 
-        HPX_CORE_EXPORT thread_domain();
+        HPX_CORE_EXPORT thread_domain() noexcept;
     };
 
     struct id
     {
-        id(domain const& domain, void* addr, unsigned long extra = 0)
+        id(domain const& domain, void* addr, unsigned long extra = 0) noexcept
         {
             id_ = HPX_ITT_MAKE_ID(addr, extra);
             HPX_ITT_ID_CREATE(domain.domain_, id_);
@@ -246,14 +253,14 @@ namespace hpx { namespace util { namespace itt {
         }
 
         id(id const& rhs) = delete;
-        id(id&& rhs)
+        id(id&& rhs) noexcept
           : id_(rhs.id_)
         {
             rhs.id_ = nullptr;
         }
 
         id& operator=(id const& rhs) = delete;
-        id& operator=(id&& rhs)
+        id& operator=(id&& rhs) noexcept
         {
             if (this != &rhs)
             {
@@ -269,15 +276,16 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct frame_context
     {
-        frame_context(domain const& domain, id* ident = nullptr)
+        frame_context(domain const& domain, id* ident = nullptr) noexcept
           : domain_(domain)
           , ident_(ident)
         {
-            HPX_ITT_FRAME_BEGIN(domain_.domain_, ident_ ? ident_->id_ : 0);
+            HPX_ITT_FRAME_BEGIN(
+                domain_.domain_, ident_ ? ident_->id_ : nullptr);
         }
         ~frame_context()
         {
-            HPX_ITT_FRAME_END(domain_.domain_, ident_ ? ident_->id_ : 0);
+            HPX_ITT_FRAME_END(domain_.domain_, ident_ ? ident_->id_ : nullptr);
         }
 
         domain const& domain_;
@@ -286,7 +294,7 @@ namespace hpx { namespace util { namespace itt {
 
     struct undo_frame_context
     {
-        undo_frame_context(frame_context& frame)
+        undo_frame_context(frame_context& frame) noexcept
           : frame_(frame)
         {
             HPX_ITT_FRAME_END(frame_.domain_.domain_, nullptr);
@@ -302,7 +310,7 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct mark_context
     {
-        mark_context(char const* name)
+        mark_context(char const* name) noexcept
           : itt_mark_(0)
           , name_(name)
         {
@@ -319,7 +327,7 @@ namespace hpx { namespace util { namespace itt {
 
     struct undo_mark_context
     {
-        undo_mark_context(mark_context& mark)
+        undo_mark_context(mark_context& mark) noexcept
           : mark_(mark)
         {
             HPX_ITT_MARK_OFF(mark_.itt_mark_);
@@ -336,11 +344,11 @@ namespace hpx { namespace util { namespace itt {
     struct string_handle
     {
         string_handle() noexcept
-          : handle_(0)
+          : handle_(nullptr)
         {
         }
-        string_handle(char const* s)
-          : handle_(s == nullptr ? 0 : HPX_ITT_STRING_HANDLE_CREATE(s))
+        string_handle(char const* s) noexcept
+          : handle_(s == nullptr ? nullptr : HPX_ITT_STRING_HANDLE_CREATE(s))
         {
         }
         string_handle(___itt_string_handle* h) noexcept
@@ -356,7 +364,7 @@ namespace hpx { namespace util { namespace itt {
 
         explicit operator bool() const noexcept
         {
-            return handle_ != 0;
+            return handle_ != nullptr;
         }
 
         ___itt_string_handle* handle_;
@@ -365,25 +373,25 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct task
     {
-        HPX_CORE_EXPORT task(
-            domain const&, string_handle const&, std::uint64_t metadata);
-        HPX_CORE_EXPORT task(domain const&, string_handle const&);
+        HPX_CORE_EXPORT task(domain const&, string_handle const&,
+            std::uint64_t metadata) noexcept;
+        HPX_CORE_EXPORT task(domain const&, string_handle const&) noexcept;
         HPX_CORE_EXPORT ~task();
 
-        void add_metadata(string_handle const& name, std::uint64_t val)
+        void add_metadata(string_handle const& name, std::uint64_t val) noexcept
         {
             HPX_ITT_METADATA_ADD(domain_.domain_, id_, name.handle_, val);
         }
-        void add_metadata(string_handle const& name, double val)
+        void add_metadata(string_handle const& name, double val) noexcept
         {
             HPX_ITT_METADATA_ADD(domain_.domain_, id_, name.handle_, val);
         }
-        void add_metadata(string_handle const& name, char const* val)
+        void add_metadata(string_handle const& name, char const* val) noexcept
         {
             HPX_ITT_METADATA_ADD(domain_.domain_, id_, name.handle_, val);
         }
         template <typename T>
-        void add_metadata(string_handle const& name, T const& val)
+        void add_metadata(string_handle const& name, T const& val) noexcept
         {
             HPX_ITT_METADATA_ADD(domain_.domain_, id_, name.handle_,
                 static_cast<void const*>(&val));
@@ -397,7 +405,7 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct heap_function
     {
-        heap_function(char const* name, char const* domain)
+        heap_function(char const* name, char const* domain) noexcept
           : heap_function_(HPX_ITT_HEAP_FUNCTION_CREATE(name, domain))
         {
         }
@@ -407,7 +415,7 @@ namespace hpx { namespace util { namespace itt {
 
     struct heap_internal_access
     {
-        heap_internal_access()
+        heap_internal_access() noexcept
         {
             HPX_ITT_HEAP_INTERNAL_ACCESS_BEGIN();
         }
@@ -421,8 +429,8 @@ namespace hpx { namespace util { namespace itt {
     struct heap_allocate
     {
         template <typename T>
-        heap_allocate(
-            heap_function& heap_function, T**& addr, std::size_t size, int init)
+        heap_allocate(heap_function& heap_function, T**& addr, std::size_t size,
+            int init) noexcept
           : heap_function_(heap_function)
           , addr_(reinterpret_cast<void**&>(addr))
           , size_(size)
@@ -447,7 +455,7 @@ namespace hpx { namespace util { namespace itt {
 
     struct heap_free
     {
-        heap_free(heap_function& heap_function, void* addr)
+        heap_free(heap_function& heap_function, void* addr) noexcept
           : heap_function_(heap_function)
           , addr_(addr)
         {
@@ -467,11 +475,11 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct counter
     {
-        counter(char const* name, char const* domain)
+        counter(char const* name, char const* domain) noexcept
           : id_(HPX_ITT_COUNTER_CREATE(name, domain))
         {
         }
-        counter(char const* name, char const* domain, int type)
+        counter(char const* name, char const* domain, int type) noexcept
           : id_(HPX_ITT_COUNTER_CREATE_TYPED(name, domain, type))
         {
         }
@@ -482,7 +490,7 @@ namespace hpx { namespace util { namespace itt {
         }
 
         template <typename T>
-        void set_value(T const& value)
+        void set_value(T const& value) noexcept
         {
             if (id_)
             {
@@ -492,14 +500,14 @@ namespace hpx { namespace util { namespace itt {
         }
 
         counter(counter const& rhs) = delete;
-        counter(counter&& rhs)
+        counter(counter&& rhs) noexcept
           : id_(rhs.id_)
         {
             rhs.id_ = nullptr;
         }
 
         counter& operator=(counter const& rhs) = delete;
-        counter& operator=(counter&& rhs)
+        counter& operator=(counter&& rhs) noexcept
         {
             if (this != &rhs)
             {
@@ -516,17 +524,17 @@ namespace hpx { namespace util { namespace itt {
     ///////////////////////////////////////////////////////////////////////////
     struct event
     {
-        event(char const* name)
+        event(char const* name) noexcept
           : event_(itt_event_create(name, (int) strnlen(name, 256)))
         {
         }
 
-        void start() const
+        void start() const noexcept
         {
             itt_event_start(event_);
         }
 
-        void end() const
+        void end() const noexcept
         {
             itt_event_end(event_);
         }
@@ -537,7 +545,7 @@ namespace hpx { namespace util { namespace itt {
 
     struct mark_event
     {
-        mark_event(event const& e)
+        mark_event(event const& e) noexcept
           : e_(e)
         {
             e_.start();
@@ -551,7 +559,7 @@ namespace hpx { namespace util { namespace itt {
         event e_;
     };
 
-    inline void event_tick(event const& e)
+    inline void event_tick(event const& e) noexcept
     {
         e.start();
     }
@@ -559,137 +567,160 @@ namespace hpx { namespace util { namespace itt {
 
 #else
 
-inline void itt_sync_create(void*, const char*, const char*) {}
-inline void itt_sync_rename(void*, const char*) {}
-inline void itt_sync_prepare(void*) {}
-inline void itt_sync_acquired(void*) {}
-inline void itt_sync_cancel(void*) {}
-inline void itt_sync_releasing(void*) {}
-inline void itt_sync_released(void*) {}
-inline void itt_sync_destroy(void*) {}
+inline constexpr void itt_sync_create(void*, const char*, const char*) noexcept
+{
+}
+inline constexpr void itt_sync_rename(void*, const char*) noexcept {}
+inline constexpr void itt_sync_prepare(void*) noexcept {}
+inline constexpr void itt_sync_acquired(void*) noexcept {}
+inline constexpr void itt_sync_cancel(void*) noexcept {}
+inline constexpr void itt_sync_releasing(void*) noexcept {}
+inline constexpr void itt_sync_released(void*) noexcept {}
+inline constexpr void itt_sync_destroy(void*) noexcept {}
 
-inline ___itt_caller* itt_stack_create()
+inline constexpr ___itt_caller* itt_stack_create() noexcept
 {
     return nullptr;
 }
-inline void itt_stack_enter(___itt_caller*) {}
-inline void itt_stack_leave(___itt_caller*) {}
-inline void itt_stack_destroy(___itt_caller*) {}
+inline constexpr void itt_stack_enter(___itt_caller*) noexcept {}
+inline constexpr void itt_stack_leave(___itt_caller*) noexcept {}
+inline constexpr void itt_stack_destroy(___itt_caller*) noexcept {}
 
-inline void itt_frame_begin(___itt_domain const*, ___itt_id*) {}
-inline void itt_frame_end(___itt_domain const*, ___itt_id*) {}
+inline constexpr void itt_frame_begin(___itt_domain const*, ___itt_id*) noexcept
+{
+}
+inline constexpr void itt_frame_end(___itt_domain const*, ___itt_id*) noexcept
+{
+}
 
-inline int itt_mark_create(char const*)
+inline constexpr int itt_mark_create(char const*) noexcept
 {
     return 0;
 }
-inline void itt_mark_off(int) {}
-inline void itt_mark(int, char const*) {}
+inline constexpr void itt_mark_off(int) noexcept {}
+inline constexpr void itt_mark(int, char const*) noexcept {}
 
-inline void itt_thread_set_name(char const*) {}
-inline void itt_thread_ignore() {}
+inline constexpr void itt_thread_set_name(char const*) noexcept {}
+inline constexpr void itt_thread_ignore() noexcept {}
 
-inline void itt_task_begin(___itt_domain const*, ___itt_string_handle*) {}
-inline void itt_task_begin(
-    ___itt_domain const*, ___itt_id*, ___itt_string_handle*)
+inline constexpr void itt_task_begin(
+    ___itt_domain const*, ___itt_string_handle*) noexcept
 {
 }
-inline void itt_task_end(___itt_domain const*) {}
+inline constexpr void itt_task_begin(
+    ___itt_domain const*, ___itt_id*, ___itt_string_handle*) noexcept
+{
+}
+inline constexpr void itt_task_end(___itt_domain const*) noexcept {}
 
-inline ___itt_domain* itt_domain_create(char const*)
+inline constexpr ___itt_domain* itt_domain_create(char const*) noexcept
 {
     return nullptr;
 }
-inline ___itt_string_handle* itt_string_handle_create(char const*)
+inline constexpr ___itt_string_handle* itt_string_handle_create(
+    char const*) noexcept
 {
     return nullptr;
 }
 
-inline ___itt_id* itt_make_id(void*, unsigned long)
+inline constexpr ___itt_id* itt_make_id(void*, unsigned long)
 {
     return nullptr;
 }
-inline void itt_id_create(___itt_domain const*, ___itt_id*) {}
-inline void itt_id_destroy(___itt_id*) {}
+inline constexpr void itt_id_create(___itt_domain const*, ___itt_id*) noexcept
+{
+}
+inline constexpr void itt_id_destroy(___itt_id*) noexcept {}
 
-inline __itt_heap_function itt_heap_function_create(const char*, const char*)
+inline constexpr __itt_heap_function itt_heap_function_create(
+    const char*, const char*) noexcept
 {
     return nullptr;
 }
-inline void itt_heap_allocate_begin(__itt_heap_function, std::size_t, int) {}
-inline void itt_heap_allocate_end(__itt_heap_function, void**, std::size_t, int)
+inline constexpr void itt_heap_allocate_begin(
+    __itt_heap_function, std::size_t, int) noexcept
 {
 }
-inline void itt_heap_free_begin(__itt_heap_function, void*) {}
-inline void itt_heap_free_end(__itt_heap_function, void*) {}
-inline void itt_heap_reallocate_begin(
-    __itt_heap_function, void*, std::size_t, int)
+inline constexpr void itt_heap_allocate_end(
+    __itt_heap_function, void**, std::size_t, int) noexcept
 {
 }
-inline void itt_heap_reallocate_end(
-    __itt_heap_function, void*, void**, std::size_t, int)
+inline constexpr void itt_heap_free_begin(__itt_heap_function, void*) noexcept
 {
 }
-inline void itt_heap_internal_access_begin() {}
-inline void itt_heap_internal_access_end() {}
+inline constexpr void itt_heap_free_end(__itt_heap_function, void*) noexcept {}
+inline constexpr void itt_heap_reallocate_begin(
+    __itt_heap_function, void*, std::size_t, int) noexcept
+{
+}
+inline constexpr void itt_heap_reallocate_end(
+    __itt_heap_function, void*, void**, std::size_t, int) noexcept
+{
+}
+inline constexpr void itt_heap_internal_access_begin() noexcept {}
+inline constexpr void itt_heap_internal_access_end() noexcept {}
 
-inline ___itt_counter* itt_counter_create(char const*, char const*)
+inline constexpr ___itt_counter* itt_counter_create(
+    char const*, char const*) noexcept
 {
     return nullptr;
 }
-inline ___itt_counter* itt_counter_create_typed(char const*, char const*, int)
+inline constexpr ___itt_counter* itt_counter_create_typed(
+    char const*, char const*, int) noexcept
 {
     return nullptr;
 }
-inline void itt_counter_destroy(___itt_counter*) {}
-inline void itt_counter_set_value(___itt_counter*, void*) {}
+inline constexpr void itt_counter_destroy(___itt_counter*) noexcept {}
+inline constexpr void itt_counter_set_value(___itt_counter*, void*) noexcept {}
 
-inline int itt_event_create(char const*, int)
+inline constexpr int itt_event_create(char const*, int) noexcept
 {
     return 0;
 }
-inline int itt_event_start(int)
+inline constexpr int itt_event_start(int) noexcept
 {
     return 0;
 }
-inline int itt_event_end(int)
+inline constexpr int itt_event_end(int) noexcept
 {
     return 0;
 }
 
-inline void itt_metadata_add(
-    ___itt_domain*, ___itt_id*, ___itt_string_handle*, std::uint64_t const&)
+inline constexpr void itt_metadata_add(___itt_domain*, ___itt_id*,
+    ___itt_string_handle*, std::uint64_t const&) noexcept
 {
 }
-inline void itt_metadata_add(
-    ___itt_domain*, ___itt_id*, ___itt_string_handle*, double const&)
+inline constexpr void itt_metadata_add(
+    ___itt_domain*, ___itt_id*, ___itt_string_handle*, double const&) noexcept
 {
 }
-inline void itt_metadata_add(
-    ___itt_domain*, ___itt_id*, ___itt_string_handle*, char const*)
+inline constexpr void itt_metadata_add(
+    ___itt_domain*, ___itt_id*, ___itt_string_handle*, char const*) noexcept
 {
 }
-inline void itt_metadata_add(
-    ___itt_domain*, ___itt_id*, ___itt_string_handle*, void const*)
+inline constexpr void itt_metadata_add(
+    ___itt_domain*, ___itt_id*, ___itt_string_handle*, void const*) noexcept
 {
 }
 
 //////////////////////////////////////////////////////////////////////////////
 namespace hpx { namespace util {
+
     struct thread_description;
 }}    // namespace hpx::util
 
 namespace hpx { namespace util { namespace itt {
+
     struct stack_context
     {
-        stack_context() {}
-        ~stack_context() {}
+        stack_context() = default;
+        ~stack_context() = default;
     };
 
     struct caller_context
     {
-        caller_context(stack_context&) {}
-        ~caller_context() {}
+        constexpr caller_context(stack_context&) noexcept {}
+        ~caller_context() = default;
     };
 
     //////////////////////////////////////////////////////////////////////////
@@ -697,114 +728,116 @@ namespace hpx { namespace util { namespace itt {
     {
         HPX_NON_COPYABLE(domain);
 
-        domain(char const*) {}
-        domain() {}
+        constexpr domain(char const*) noexcept {}
+        domain() = default;
     };
 
     struct thread_domain : domain
     {
         HPX_NON_COPYABLE(thread_domain);
 
-        thread_domain()
-          : domain(){};
+        thread_domain() = default;
     };
 
     struct id
     {
-        id(domain const& /*domain*/, void* /*addr*/,
-            unsigned long /*extra*/ = 0)
-        {
-        }
-        ~id() {}
+        constexpr id(domain const&, void*, unsigned long = 0) noexcept {}
+        ~id() = default;
     };
 
     ///////////////////////////////////////////////////////////////////////////
     struct frame_context
     {
-        frame_context(domain const&, id* = nullptr) {}
-        ~frame_context() {}
+        constexpr frame_context(domain const&, id* = nullptr) noexcept {}
+        ~frame_context() = default;
     };
 
     struct undo_frame_context
     {
-        undo_frame_context(frame_context const&) {}
-        ~undo_frame_context() {}
+        constexpr undo_frame_context(frame_context const&) noexcept {}
+        ~undo_frame_context() = default;
     };
 
     ///////////////////////////////////////////////////////////////////////////
     struct mark_context
     {
-        mark_context(char const*) {}
-        ~mark_context() {}
+        constexpr mark_context(char const*) noexcept {}
+        ~mark_context() = default;
     };
 
     struct undo_mark_context
     {
-        undo_mark_context(mark_context const&) {}
-        ~undo_mark_context() {}
+        constexpr undo_mark_context(mark_context const&) noexcept {}
+        ~undo_mark_context() = default;
     };
 
     ///////////////////////////////////////////////////////////////////////////
     struct string_handle
     {
-        string_handle(char const* = nullptr) {}
+        constexpr string_handle(char const* = nullptr) noexcept {}
     };
 
     //////////////////////////////////////////////////////////////////////////
     struct task
     {
-        task(domain const&, string_handle const&, std::uint64_t) {}
-        task(domain const&, string_handle const&) {}
+        constexpr task(
+            domain const&, string_handle const&, std::uint64_t) noexcept
+        {
+        }
+        constexpr task(domain const&, string_handle const&) noexcept {}
 
-        ~task() {}
+        ~task() = default;
     };
 
     ///////////////////////////////////////////////////////////////////////////
     struct heap_function
     {
-        heap_function(char const*, char const*) {}
-        ~heap_function() {}
+        constexpr heap_function(char const*, char const*) noexcept {}
+        ~heap_function() = default;
     };
 
     struct heap_allocate
     {
         template <typename T>
-        heap_allocate(heap_function& /*heap_function*/, T**, std::size_t, int)
+        constexpr heap_allocate(
+            heap_function& /*heap_function*/, T**, std::size_t, int) noexcept
         {
         }
-        ~heap_allocate() {}
+        ~heap_allocate() = default;
     };
 
     struct heap_free
     {
-        heap_free(heap_function& /*heap_function*/, void*) {}
-        ~heap_free() {}
+        constexpr heap_free(heap_function& /*heap_function*/, void*) noexcept {}
+        ~heap_free() = default;
     };
 
     struct heap_internal_access
     {
-        heap_internal_access() {}
-        ~heap_internal_access() {}
+        heap_internal_access() = default;
+        ~heap_internal_access() = default;
     };
 
     struct counter
     {
-        counter(char const* /*name*/, char const* /*domain*/) {}
-        ~counter() {}
+        constexpr counter(char const* /*name*/, char const* /*domain*/) noexcept
+        {
+        }
+        ~counter() = default;
     };
 
     struct event
     {
-        event(char const*) {}
+        constexpr event(char const*) noexcept {}
     };
 
     struct mark_event
     {
-        mark_event(event const&) {}
-        ~mark_event() {}
+        constexpr mark_event(event const&) noexcept {}
+        ~mark_event() = default;
     };
 
-    inline void event_tick(event const&) {}
+    inline constexpr void event_tick(event const&) noexcept {}
 }}}    // namespace hpx::util::itt
 
 #endif    // HPX_HAVE_ITTNOTIFY

--- a/libs/core/itt_notify/src/itt_notify.cpp
+++ b/libs/core/itt_notify/src/itt_notify.cpp
@@ -1,4 +1,4 @@
-//  Copyright (c) 2007-2017 Hartmut Kaiser
+//  Copyright (c) 2007-2021 Hartmut Kaiser
 //
 //  SPDX-License-Identifier: BSL-1.0
 //  Distributed under the Boost Software License, Version 1.0. (See accompanying
@@ -247,7 +247,8 @@ bool use_ittnotify_api = false;
 
 ///////////////////////////////////////////////////////////////////////////////
 namespace hpx { namespace util { namespace itt {
-    domain::domain(char const* name)
+
+    domain::domain(char const* name) noexcept
       : domain_(HPX_ITT_DOMAIN_CREATE(name))
     {
         if (domain_ != nullptr)
@@ -256,15 +257,15 @@ namespace hpx { namespace util { namespace itt {
         }
     }
 
-    domain::domain()
+    domain::domain() noexcept
       : domain_(nullptr)
     {
     }
 
     static thread_local std::unique_ptr<___itt_domain> thread_domain_;
 
-    thread_domain::thread_domain()
-      : domain_()
+    thread_domain::thread_domain() noexcept
+      : domain()
     {
         if (thread_domain_.get() == nullptr)
         {
@@ -280,9 +281,9 @@ namespace hpx { namespace util { namespace itt {
         }
     }
 
-    task::task(domain const& domain, string_handle const& name)
+    task::task(domain const& domain, string_handle const& name) noexcept
       : domain_(domain)
-      , id_(0)
+      , id_(nullptr)
       , sh_(name)
     {
         id_ = HPX_ITT_MAKE_ID(
@@ -291,8 +292,8 @@ namespace hpx { namespace util { namespace itt {
         HPX_ITT_TASK_BEGIN_ID(domain_.domain_, id_, sh_.handle_);
     }
 
-    task::task(
-        domain const& domain, string_handle const& name, std::uint64_t metadata)
+    task::task(domain const& domain, string_handle const& name,
+        std::uint64_t metadata) noexcept
       : domain_(domain)
       , id_(0)
       , sh_(name)
@@ -313,128 +314,130 @@ namespace hpx { namespace util { namespace itt {
 }}}    // namespace hpx::util::itt
 
 ///////////////////////////////////////////////////////////////////////////////
-void itt_sync_create(void* addr, const char* objtype, const char* objname)
+void itt_sync_create(
+    void* addr, const char* objtype, const char* objname) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_CREATE(addr, objtype, objname);
 }
 
-void itt_sync_rename(void* addr, const char* objname)
+void itt_sync_rename(void* addr, const char* objname) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_RENAME(addr, objname);
 }
 
-void itt_sync_prepare(void* addr)
+void itt_sync_prepare(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_PREPARE(addr);
 }
 
-void itt_sync_acquired(void* addr)
+void itt_sync_acquired(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_ACQUIRED(addr);
 }
 
-void itt_sync_cancel(void* addr)
+void itt_sync_cancel(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_CANCEL(addr);
 }
 
-void itt_sync_releasing(void* addr)
+void itt_sync_releasing(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_RELEASING(addr);
 }
 
-void itt_sync_released(void* addr)
+void itt_sync_released(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_RELEASED(addr);
 }
 
-void itt_sync_destroy(void* addr)
+void itt_sync_destroy(void* addr) noexcept
 {
     HPX_INTERNAL_ITT_SYNC_DESTROY(addr);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-__itt_caller itt_stack_create()
+__itt_caller itt_stack_create() noexcept
 {
     return HPX_INTERNAL_ITT_STACK_CREATE();
 }
 
-void itt_stack_enter(__itt_caller ctx)
+void itt_stack_enter(__itt_caller ctx) noexcept
 {
     HPX_INTERNAL_ITT_STACK_ENTER(ctx);
 }
 
-void itt_stack_leave(__itt_caller ctx)
+void itt_stack_leave(__itt_caller ctx) noexcept
 {
     HPX_INTERNAL_ITT_STACK_LEAVE(ctx);
 }
 
-void itt_stack_destroy(__itt_caller ctx)
+void itt_stack_destroy(__itt_caller ctx) noexcept
 {
     HPX_INTERNAL_ITT_STACK_DESTROY(ctx);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-void itt_frame_begin(___itt_domain const* domain, ___itt_id* id)
+void itt_frame_begin(___itt_domain const* domain, ___itt_id* id) noexcept
 {
     HPX_INTERNAL_ITT_FRAME_BEGIN(domain, id);
 }
 
-void itt_frame_end(___itt_domain const* domain, ___itt_id* id)
+void itt_frame_end(___itt_domain const* domain, ___itt_id* id) noexcept
 {
     HPX_INTERNAL_ITT_FRAME_END(domain, id);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-int itt_mark_create(char const* name)
+int itt_mark_create(char const* name) noexcept
 {
     return HPX_INTERNAL_ITT_MARK_CREATE(name);
 }
 
-void itt_mark_off(int mark)
+void itt_mark_off(int mark) noexcept
 {
     HPX_INTERNAL_ITT_MARK_OFF(mark);
 }
 
-void itt_mark(int mark, char const* par)
+void itt_mark(int mark, char const* par) noexcept
 {
     HPX_INTERNAL_ITT_MARK(mark, par);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-void itt_thread_set_name(char const* name)
+void itt_thread_set_name(char const* name) noexcept
 {
     HPX_INTERNAL_ITT_THREAD_SET_NAME(name);
 }
 
-void itt_thread_ignore()
+void itt_thread_ignore() noexcept
 {
     HPX_INTERNAL_ITT_THREAD_IGNORE();
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-void itt_task_begin(___itt_domain const* domain, ___itt_string_handle* name)
+void itt_task_begin(
+    ___itt_domain const* domain, ___itt_string_handle* name) noexcept
 {
     HPX_INTERNAL_ITT_TASK_BEGIN(domain, name);
 }
 
-void itt_task_begin(
-    ___itt_domain const* domain, ___itt_id* id, ___itt_string_handle* name)
+void itt_task_begin(___itt_domain const* domain, ___itt_id* id,
+    ___itt_string_handle* name) noexcept
 {
     HPX_INTERNAL_ITT_TASK_BEGIN_ID(domain, *id, name);
 }
 
-void itt_task_end(___itt_domain const* domain)
+void itt_task_end(___itt_domain const* domain) noexcept
 {
     HPX_INTERNAL_ITT_TASK_END(domain);
 }
 
-___itt_domain* itt_domain_create(char const* name)
+___itt_domain* itt_domain_create(char const* name) noexcept
 {
     return HPX_INTERNAL_ITT_DOMAIN_CREATE(name);
 }
 
-___itt_string_handle* itt_string_handle_create(char const* name)
+___itt_string_handle* itt_string_handle_create(char const* name) noexcept
 {
     return HPX_INTERNAL_ITT_STRING_HANDLE_CREATE(name);
 }
@@ -444,128 +447,129 @@ ___itt_id* itt_make_id(void* addr, std::size_t extra)
     return new ___itt_id(HPX_INTERNAL_ITT_MAKE_ID(addr, extra));
 }
 
-void itt_id_create(___itt_domain const* domain, ___itt_id* id)
+void itt_id_create(___itt_domain const* domain, ___itt_id* id) noexcept
 {
     HPX_INTERNAL_ITT_ID_CREATE(domain, *id);
 }
 
-void itt_id_destroy(___itt_id* id)
+void itt_id_destroy(___itt_id* id) noexcept
 {
     HPX_INTERNAL_ITT_ID_DESTROY(id);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 __itt_heap_function itt_heap_function_create(
-    const char* name, const char* domain)
+    const char* name, const char* domain) noexcept
 {
     return HPX_INTERNAL_ITT_HEAP_FUNCTION_CREATE(name, domain);
 }
 
-void itt_heap_allocate_begin(__itt_heap_function f, std::size_t size, int init)
+void itt_heap_allocate_begin(
+    __itt_heap_function f, std::size_t size, int init) noexcept
 {
     HPX_INTERNAL_HEAP_ALLOCATE_BEGIN(f, size, init);
 }
 
 void itt_heap_allocate_end(
-    __itt_heap_function f, void** addr, std::size_t size, int init)
+    __itt_heap_function f, void** addr, std::size_t size, int init) noexcept
 {
     HPX_INTERNAL_HEAP_ALLOCATE_END(f, addr, size, init);
 }
 
-void itt_heap_free_begin(__itt_heap_function f, void* addr)
+void itt_heap_free_begin(__itt_heap_function f, void* addr) noexcept
 {
     HPX_INTERNAL_HEAP_FREE_BEGIN(f, addr);
 }
 
-void itt_heap_free_end(__itt_heap_function f, void* addr)
+void itt_heap_free_end(__itt_heap_function f, void* addr) noexcept
 {
     HPX_INTERNAL_HEAP_FREE_END(f, addr);
 }
 
 void itt_heap_reallocate_begin(
-    __itt_heap_function f, void* addr, std::size_t new_size, int init)
+    __itt_heap_function f, void* addr, std::size_t new_size, int init) noexcept
 {
     HPX_INTERNAL_HEAP_REALLOCATE_BEGIN(f, addr, new_size, init);
 }
 
 void itt_heap_reallocate_end(__itt_heap_function f, void* addr, void** new_addr,
-    std::size_t new_size, int init)
+    std::size_t new_size, int init) noexcept
 {
     HPX_INTERNAL_HEAP_REALLOCATE_END(f, addr, new_addr, new_size, init);
 }
 
-void itt_heap_internal_access_begin()
+void itt_heap_internal_access_begin() noexcept
 {
     HPX_INTERNAL_INTERNAL_ACCESS_BEGIN();
 }
 
-void itt_heap_internal_access_end()
+void itt_heap_internal_access_end() noexcept
 {
     HPX_INTERNAL_INTERNAL_ACCESS_END();
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-__itt_counter itt_counter_create(char const* name, char const* domain)
+__itt_counter itt_counter_create(char const* name, char const* domain) noexcept
 {
     return HPX_INTERNAL_COUNTER_CREATE(name, domain);
 }
 
 __itt_counter itt_counter_create_typed(
-    char const* name, char const* domain, int type)
+    char const* name, char const* domain, int type) noexcept
 {
     return HPX_INTERNAL_COUNTER_CREATE_TYPED(
         name, domain, (__itt_metadata_type) type);
 }
 
-void itt_counter_destroy(__itt_counter id)
+void itt_counter_destroy(__itt_counter id) noexcept
 {
     HPX_INTERNAL_COUNTER_DESTROY(id);
 }
 
-void itt_counter_set_value(__itt_counter id, void* value_ptr)
+void itt_counter_set_value(__itt_counter id, void* value_ptr) noexcept
 {
     HPX_INTERNAL_COUNTER_SET_VALUE(id, value_ptr);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-__itt_event itt_event_create(char const* name, int namelen)
+__itt_event itt_event_create(char const* name, int namelen) noexcept
 {
     return HPX_INTERNAL_EVENT_CREATE(name, namelen);
 }
 
-int itt_event_start(__itt_event evnt)
+int itt_event_start(__itt_event evnt) noexcept
 {
     return HPX_INTERNAL_EVENT_START(evnt);
 }
 
-int itt_event_end(__itt_event evnt)
+int itt_event_end(__itt_event evnt) noexcept
 {
     return HPX_INTERNAL_EVENT_END(evnt);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, std::uint64_t const& data)
+    ___itt_string_handle* key, std::uint64_t const& data) noexcept
 {
     HPX_INTERNAL_METADATA_ADD(domain, *id, key, __itt_metadata_u64, 1,
         const_cast<std::uint64_t*>(&data));
 }
 
 void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, double const& data)
+    ___itt_string_handle* key, double const& data) noexcept
 {
     HPX_INTERNAL_METADATA_ADD(
         domain, *id, key, __itt_metadata_double, 1, const_cast<double*>(&data));
 }
 
 void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, char const* data)
+    ___itt_string_handle* key, char const* data) noexcept
 {
     HPX_INTERNAL_METADATA_STR_ADD(domain, *id, key, data);
 }
 
 void itt_metadata_add(___itt_domain* domain, ___itt_id* id,
-    ___itt_string_handle* key, void const* data)
+    ___itt_string_handle* key, void const* data) noexcept
 {
     HPX_INTERNAL_METADATA_ADD(
         domain, *id, key, __itt_metadata_unknown, 1, const_cast<void*>(data));

--- a/libs/core/synchronization/include/hpx/synchronization/spinlock_pool.hpp
+++ b/libs/core/synchronization/include/hpx/synchronization/spinlock_pool.hpp
@@ -1,4 +1,4 @@
-//  Copyright (c) 2012 Hartmut Kaiser
+//  Copyright (c) 2012-2021 Hartmut Kaiser
 //  Copyright (c) 2014 Thomas Heller
 //
 //  adapted from:
@@ -22,16 +22,17 @@
 
 ///////////////////////////////////////////////////////////////////////////////
 namespace hpx { namespace lcos { namespace local {
-    namespace detail {
+
 #if HPX_HAVE_ITTNOTIFY != 0
+    namespace detail {
         template <typename Tag, std::size_t N>
         struct itt_spinlock_init
         {
-            itt_spinlock_init();
+            itt_spinlock_init() noexcept;
             ~itt_spinlock_init();
         };
-#endif
     }    // namespace detail
+#endif
 
     template <typename Tag, std::size_t N = HPX_HAVE_SPINLOCK_POOL_NUM>
     class spinlock_pool
@@ -39,10 +40,10 @@ namespace hpx { namespace lcos { namespace local {
     private:
         static util::cache_aligned_data<lcos::local::spinlock> pool_[N];
 #if HPX_HAVE_ITTNOTIFY != 0
-        static detail::itt_spinlock_init<Tag> init_;
+        static detail::itt_spinlock_init<Tag, N> init_;
 #endif
     public:
-        static lcos::local::spinlock& spinlock_for(void const* pv)
+        static lcos::local::spinlock& spinlock_for(void const* pv) noexcept
         {
             std::size_t i = util::fibhash<N>(reinterpret_cast<std::size_t>(pv));
             return pool_[i].data_;
@@ -90,16 +91,17 @@ namespace hpx { namespace lcos { namespace local {
 
 #if HPX_HAVE_ITTNOTIFY != 0
     namespace detail {
+
         template <typename Tag, std::size_t N>
-        itt_spinlock_init<Tag, N>::itt_spinlock_init()
+        itt_spinlock_init<Tag, N>::itt_spinlock_init() noexcept
         {
-            for (int i = 0; i < 41; ++i)
+            for (int i = 0; i < N; ++i)
             {
                 HPX_ITT_SYNC_CREATE(
-                    &lcos::local::spinlock_pool<Tag, N>::pool_[i].data_,
+                    (&lcos::local::spinlock_pool<Tag, N>::pool_[i].data_),
                     "hpx::lcos::spinlock", 0);
                 HPX_ITT_SYNC_RENAME(
-                    &lcos::local::spinlock_pool<Tag, N>::pool_[i].data_,
+                    (&lcos::local::spinlock_pool<Tag, N>::pool_[i].data_),
                     "hpx::lcos::spinlock");
             }
         }
@@ -109,12 +111,12 @@ namespace hpx { namespace lcos { namespace local {
         {
             for (int i = 0; i < N; ++i)
             {
-                HPX_ITT_SYNC_DESTROY(&spinlock_pool<Tag, N>::pool_[i].data_);
+                HPX_ITT_SYNC_DESTROY((&spinlock_pool<Tag, N>::pool_[i].data_));
             }
         }
     }    // namespace detail
 
     template <typename Tag, std::size_t N>
-    util::detail::itt_spinlock_init<Tag, N> spinlock_pool<Tag, N>::init_;
+    detail::itt_spinlock_init<Tag, N> spinlock_pool<Tag, N>::init_{};
 #endif
 }}}    // namespace hpx::lcos::local

--- a/libs/full/actions_base/include/hpx/actions_base/basic_action.hpp
+++ b/libs/full/actions_base/include/hpx/actions_base/basic_action.hpp
@@ -46,7 +46,7 @@
 #include <hpx/type_support/pack.hpp>
 #include <hpx/util/get_and_reset_value.hpp>
 #if HPX_HAVE_ITTNOTIFY != 0 && !defined(HPX_HAVE_APEX)
-#include <hpx/concurrency/itt_notify.hpp>
+#include <hpx/modules/itt_notify.hpp>
 #endif
 
 #include <boost/utility/string_ref.hpp>

--- a/libs/full/components_base/src/server/wrapper_heap.cpp
+++ b/libs/full/components_base/src/server/wrapper_heap.cpp
@@ -89,6 +89,7 @@ namespace hpx { namespace components { namespace detail {
       , heap_free_function_("wrapper_heap::free", class_name)
     {
         util::itt::heap_internal_access hia;
+        HPX_UNUSED(hia);
         if (!init_pool())
             throw std::bad_alloc();
     }

